src/main/java/com/blamejared/crafttweaker/impl/recipes/replacement/ReplacerAction.java

```diff
@@ -81,6 +85,20 @@ public boolean validate(final ILogger logger) {
         return true;
     }
     
+    private String joinBothLists(final Collection<ResourceLocation> userExclusions, final Collection<ResourceLocation> defaultExclusions) {
+        if (userExclusions.isEmpty() && defaultExclusions.isEmpty()) return "";
+        
+        final List<String> userExcludedNames = userExclusions.stream().map(ResourceLocation::toString).collect(Collectors.toList());
+        final List<String> defaultExcludedNames = defaultExclusions.stream().map(ResourceLocation::toString).collect(Collectors.toList());
+        
+        return String.format(
+                " excluding {%s%s%s}",
+                userExcludedNames.isEmpty()? "" : userExcludedNames.stream().collect(Collectors.joining("\", \"", "\"", "\"")),
+                userExcludedNames.isEmpty() || defaultExcludedNames.isEmpty()? "" : ", ",
+                defaultExcludedNames.isEmpty()? "" : defaultExcludedNames.stream().collect(Collectors.joining("\" (automatic), \"", "\"", "\" (automatic)"))
```

Comment:
I am not particularly fond of `(automatic)` as a marker. Does anyone have a better idea?